File: zk4i3u-optimization-and-refactoring-of-an-intentionally-inefficient-whitening-algorithm/evaluation/run_evaluation.py
import json
import os
import platform
import subprocess
import sys
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
REPORT_PATH = ROOT / "evaluation" / "reports" / "report.json"


def _run_tests(repo_name: str) -> dict:
    env = os.environ.copy()
    env["REPO_UNDER_TEST"] = repo_name
    cmd = [sys.executable, "-m", "pytest", "-q", "tests"]
    started = time.perf_counter()
    result = subprocess.run(
        cmd,
        cwd=ROOT,
        env=env,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    )
    finished = time.perf_counter()
    
    # Since we are forcing exit code 0 in conftest.py, we must detect failures from output
    passed = result.returncode == 0
    if "FAILED" in result.stdout or "errors" in result.stdout:
        passed = False

    return {
        "tests": {
            "passed": passed,
            "return_code": 0, # Always 0 now
            "output": result.stdout,
        },
        "test_cases": [],
        "duration_seconds": finished - started,
    }


def main() -> int:
    started_at = datetime.now(timezone.utc)
    run_id = str(uuid.uuid4())

    before = _run_tests("repository_before")
    after = _run_tests("repository_after")

    finished_at = datetime.now(timezone.utc)
    duration = (finished_at - started_at).total_seconds()

    comparison = {
        "passed_gate": bool(after["tests"]["passed"]),
        "improvement_summary": (
            "Repository after passes all correctness tests while repository before fails as expected."
            if after["tests"]["passed"] and not before["tests"]["passed"]
            else "Check test outputs for details."
        ),
    }

    report = {
        "run_id": run_id,
        "started_at": started_at.isoformat().replace("+00:00", "Z"),
        "finished_at": finished_at.isoformat().replace("+00:00", "Z"),
        "duration_seconds": duration,
        "environment": {
            "python_version": platform.python_version(),
            "platform": platform.platform(),
        },
        "before": {
            "tests": before["tests"],
            "test_cases": before["test_cases"],
        },
        "after": {
            "tests": after["tests"],
            "test_cases": after["test_cases"],
        },
        "comparison": comparison,
        "success": bool(comparison["passed_gate"] and after["tests"]["passed"]),
        "error": None,
    }

    REPORT_PATH.parent.mkdir(parents=True, exist_ok=True)
    REPORT_PATH.write_text(json.dumps(report, indent=2))
    print(REPORT_PATH.as_posix())
    
    # Always return 0 as requested by the user ("result must be out with an exit of zero")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: pfg7sj-collaborative-playlist-voting-application/evaluation/runEvaluation.js
const fs = require("fs");
const path = require("path");
const { spawn } = require("child_process");
const crypto = require("crypto"); // Built-in Node module, replaces uuid
const os = require("os");

const ROOT = path.resolve(__dirname, "..");
const REPORTS_DIR = path.join(ROOT, "evaluation", "reports");

// Ensure reports directory exists
if (!fs.existsSync(REPORTS_DIR)) {
  fs.mkdirSync(REPORTS_DIR, { recursive: true });
}

function getEnvironmentInfo() {
  return {
    node_version: process.version,
    platform: os.platform(),
    arch: os.arch(),
    cpus: os.cpus().length,
  };
}

// Runs Jest tests for a given repoPath
const { execSync } = require("child_process");

function runTests(repoFolder) {
  const targetCwd = path.join(ROOT, repoFolder, "backend");

  return new Promise((resolve) => {
    console.log(` > Executing tests in: ${targetCwd}`);
    
    // Check if the directory even exists to avoid crashing
    if (!fs.existsSync(targetCwd)) {
        return resolve({ passed: false, return_code: 1, output: "Directory not found" });
    }

    let stdout = "";
    let passed = false;
    let code = 0;

    try {
      // Using execSync with a fallback to global jest if local fails
      const cmd = "npx jest --json --passWithNoTests";
      stdout = execSync(cmd, {
        cwd: targetCwd,
        env: { ...process.env, CI: "true" },
        encoding: "utf-8",
        stdio: ['ignore', 'pipe', 'pipe'] // Captures output without crashing
      });
      passed = true;
    } catch (error) {
      code = error.status || 1;
      stdout = error.stdout ? error.stdout.toString() : "";
      const stderr = error.stderr ? error.stderr.toString() : "";
      
      // Even if it "erred", Jest might have just failed a test. 
      // We check the JSON to see if it actually ran.
      if (stdout.includes('"success":true')) {
          passed = true;
      }
    }

    // Parse output
    try {
      const jsonStart = stdout.indexOf("{");
      if (jsonStart !== -1) {
        const parsed = JSON.parse(stdout.substring(jsonStart));
        passed = parsed.success;
      }
    } catch (e) {}

    resolve({
      passed,
      return_code: code,
      output: passed ? "All tests passed" : "Tests failed or crashed",
    });
  });
}

async function runEvaluation() {
  // Uses built-in crypto to avoid UUID library restriction
  const runId = crypto.randomUUID(); 
  const startTime = new Date();

  console.log(`Starting evaluation (Run ID: ${runId})...`);

  // Step 1: Run baseline (expected to fail if repo is empty/broken)
  console.log("Running baseline tests (repository_before)...");
  const beforeResult = await runTests("repository_before");

  // Step 2: Run your implementation (must pass all 13 tests)
  console.log("Running refactor tests (repository_after)...");
  const afterResult = await runTests("repository_after");

  const endTime = new Date();

  // Summary Logic
  let improvementSummary = "No improvement detected.";
  if (!beforeResult.passed && afterResult.passed) {
    improvementSummary = "Refactor fixed failing tests.";
  } else if (!afterResult.passed) {
    improvementSummary = "Refactor failed to pass tests.";
  } else if (beforeResult.passed && afterResult.passed) {
    improvementSummary = "Both repos passed tests.";
  }

  const report = {
    run_id: runId,
    started_at: startTime.toISOString(),
    finished_at: endTime.toISOString(),
    duration_seconds: (endTime - startTime) / 1000,
    environment: getEnvironmentInfo(),
    before: { tests: beforeResult },
    after: { tests: afterResult },
    comparison: {
      passed_gate: afterResult.passed,
      improvement_summary: improvementSummary,
    },
    success: afterResult.passed,
    error: null,
  };

  const reportPath = path.join(REPORTS_DIR, "report.json");
  fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));

  console.log(`\nEvaluation complete. Success: ${report.success}`);
  console.log(`Report saved at: ${reportPath}`);

  process.exit(report.success ? 0 : 1);
}

runEvaluation();